File: library/java/ds/rmq/RangeAddRangeMin.java
package ds.rmq;

import static java.lang.Math.*;

/**
 * 区間へのaddと、区間のmin
 */
public class RangeAddRangeMin {

	final static int INF = 1001001001;
	int n;
	long[] segMin;
	long[] segAdd;

	RangeAddRangeMin(int n) {
		this.n = n;
		int size = n <= 1 ? 1 : (Integer.highestOneBit(n - 1) << 2);
		segMin = new long[size];
		segAdd = new long[size];
	}

	/**
	 * 区間[a, b)に値xを加算する。
	 */
	void add(int a, int b, int x) {
		add(a, b, x, 0, 0, n);
	}
	private void add(int a, int b, int x, int k, int l, int r)
	{
		// ノード番号 k が、範囲 [l,r)。
		if (r <= a || b <= l) return; //もし交差しない区間であれば終える.
		if (a <= l && r <= b){ //もし今みている区間[l, r)が[a, b)に完全に内包されていれば
			segAdd[k] += x;  //区間[l, r)にkを加算する.
			while (k > 0){
				//親の区間の最小値は, 子の区間の最小値 + 自分に一様に加算されている値 である.一様に加算される値は更新しなくて良い.
				k = (k - 1) / 2;
				segMin[k] = min(segMin[k * 2 + 1] + segAdd[k * 2 + 1],
								segMin[k * 2 + 2] + segAdd[k * 2 + 2]);
			}
			return;
		}
		add(a, b, x, k * 2 + 1, l, (l + r) / 2);
		add(a, b, x, k * 2 + 2, (l + r) / 2, r);
	}

	/**
	 * 区間[a, b) の最小値を返す。
	 */
	long getMin(int a, int b) {
		return getMin(a, b, 0, 0, n);
	}

	private long getMin(int a, int b, int k, int l, int r)
	{
		// ノード番号 k が、範囲 [l,r)。
		if (r <= a || b <= l) return INF;
		if (a <= l && r <= b) return (segMin[k] + segAdd[k]); //完全に内包されていれば,その区間の最小値を返す.
		long left = getMin(a, b, k * 2 + 1, l, (l + r) / 2); //子の区間の最小値を求める.
		long right = getMin(a, b, k * 2 + 2, (l + r) / 2, r); //子の区間の最小値を求める
		return min(left, right) + segAdd[k]; //親の区間の最小値は, 子の区間の最小値 + 自分に一様に加算されている値 である (大切なので2回書きました!!)
	}
}

File: library/cpp/segtree/range_add_range_min.cpp
// いわゆる Starry Sky Tree
// その区間に一様に足されている値を持つ実装
//
// [               1              ]
// [       2      ][       3      ]
// [   4  ][   5  ][   6  ][   7  ]
// [ 8][ 9][10][11][12][13][14][15]
// - root が 1 の完全二分木
// - ある頂点の番号を k とすると左の子は 2*k, 右の子は 2*k+1, 親は k/2
// - 最下段の左端の番号 == 最下段の長さ == n
const ll INF = 1LL << 61;
struct RangeAddRangeMin {
  using Val = ll;
  int n;
  vector<Val> segMin;
  vector<Val> segAdd; // 区間に一様に足されている値

  void clear(int N) {
    for (n = 1; n < N; n <<= 1);
    segMin.assign(n * 2, 0);
    segAdd.assign(n * 2, 0);
    segAdd[1] = INF; // 初期値
  }

  // [l, r) の最小値
  Val getMin(int l, int r) {
    if (r - l <= 0) return INF;
    return getMin(0, n, 1, l, r);
  }
  Val getMin(int nl, int nr, int k, int l, int r) {
    if (r <= nl || nr <= l) return INF;
    if (l <= nl && nr <= r) return segMin[k] + segAdd[k];
    Val x = getMin(nl, (nl+nr)/2, 2*k  , l, r);
    Val y = getMin((nl+nr)/2, nr, 2*k+1, l, r);
    return min(x, y) + segAdd[k];
  }

  // [l, r) の全てに val を加算
  void addRange(int l, int r, Val val) {
    addRange(0, n, 1, l, r, val);
  }
  void addRange(int nl, int nr, int k, int l, int r, Val val) {
    if (r <= nl || nr <= l) return;
    if (l <= nl && nr <= r) { segAdd[k] += val; return; }
    addRange(nl, (nl+nr)/2, 2*k  , l, r, val);
    addRange((nl+nr)/2, nr, 2*k+1, l, r, val);
    segMin[k] = min(segMin[2*k]+segAdd[2*k], segMin[2*k+1]+segAdd[2*k+1]);
  }

  // pos の値を強制的に val にする
  void update(int pos, Val val) {
    Val now = getMin(pos, pos + 1);
    addRange(pos, pos+1, val - now);
  }
};
